File: app/src/main/java/com/flyme/auto/plugin/systemui/StatusBarPlugin.java
package com.flyme.auto.plugin.systemui;

import android.service.notification.StatusBarNotification;
import android.view.View;

import com.flyme.plugin.Plugin;
import com.flyme.plugin.annotations.ProvidesInterface;

@ProvidesInterface(action = StatusBarPlugin.ACTION, version = StatusBarPlugin.VERSION)
public interface StatusBarPlugin extends Plugin {
    String ACTION = "com.flyme.auto.plugin.action.PLUGIN_STATUS_BAR";
    int VERSION = 1;

    int getDialogHeight();

    String getDialogTitle();

    View getDialogView();

    int getDialogWidth();

    @Override
    int getID();

    default void onDialogDismissed() {
    }

    default void onDialogShowed() {
    }

    default void onStatusIconPosted(StatusBarNotification sbn) {
    }

    default void setDialogCallback(DialogCallback callback) {
    }

    interface DialogCallback {
        default void dismissDialog() {
        }

        default void updateDialogTitle(String title) {
        }

        default boolean updateHeightFromAnim(int height) {
            return false;
        }

        default boolean isPanelExpended() {
            return false;
        }
    }
}

File: app/src/main/kotlin/com/flyme/auto/plugin/systemui/StatusBarPlugin.kt
package com.flyme.auto.plugin.systemui

import android.service.notification.StatusBarNotification
import android.view.View
import com.flyme.plugin.Plugin
import com.flyme.plugin.annotations.ProvidesInterface

@ProvidesInterface(action = StatusBarPlugin.ACTION, version = StatusBarPlugin.VERSION)
interface StatusBarPlugin : Plugin {
    fun getDialogHeight(): Int

    fun getDialogTitle(): String?

    fun getDialogView(): View

    fun getDialogWidth(): Int

    override fun getID(): Int

    fun onDialogDismissed() = Unit

    fun onDialogShowed() = Unit

    fun onStatusIconPosted(sbn: StatusBarNotification) = Unit

    fun setDialogCallback(callback: DialogCallback?) = Unit

    interface DialogCallback {
        fun dismissDialog() = Unit
        fun updateDialogTitle(title: String?) = Unit
        fun updateHeightFromAnim(height: Int): Boolean = false
        fun isPanelExpended(): Boolean = false
    }

    companion object {
        const val ACTION: String = "com.flyme.auto.plugin.action.PLUGIN_STATUS_BAR"
        const val VERSION: Int = 1
    }
}
